mkosi

```diff
@@ -912,12 +915,105 @@ SigLevel    = Required DatabaseOptional
 
     enable_networkd(workspace)
 
+@complete_step('Installing openSUSE')
+def install_opensuse(args, workspace, run_build_script):
+    opensuse_main_repository_map = {
+        "tumbleweed" : "http://download.opensuse.org/%s/repo/oss/",
+        "42"         : "http://download.opensuse.org/distribution/leap/%s/repo/oss/",
+        "13"         : "http://download.opensuse.org/distribution/%s/repo/oss/",
+    }
+
+    opensuse_update_repository_map = {
+        "tumbleweed" : "http://download.opensuse.org/update/%s/",
+        "42"         : "http://download.opensuse.org/update/leap/%s/oss/",
+        "13"         : "http://download.opensuse.org/update/%s/",
+    }
+
+    root = os.path.join(workspace, "root")
+    release = args.release.strip('"')
+    release = "tumbleweed" if release.isdigit() else release
+
+    try:
+        k = release.split('.')[0] if '.' in release else release
+        release_repo = ("Main", opensuse_main_repository_map[k] % release)
+        updates_repo = ("Updates", opensuse_update_repository_map[k] % release)
+    except KeyError:
+        die("Unkown openSUSE release.")
+
+    #
+    # Configure the repositories: we need to enable packages caching
+    # here to make sure that the package cache stays populated after
+    # "zypper install".
+    #
+    for (name, url) in (release_repo, updates_repo):
+        subprocess.run(["zypper", "--root", root,
+                        "addrepo", "--keep-packages",
+                        url, name], check=True)
+
+    if not args.with_docs:
+        with open(os.path.join(root, "etc/zypp/zypp.conf"), "w") as f:
+            f.write("rpm.install.excludedocs = yes\n")
+
+    # The common part of the install comand.
+    cmdline = ["zypper", "--root", root,
+                "--gpg-auto-import-keys",
+               "install", "-y", "--no-recommends"]
+    #
+    # Install the "minimal" package set.
+    #
+    subprocess.run(cmdline + ["-t", "pattern", "minimal_base"], check=True)
+
+    #
+    # Now install the additional packages if necessary.
+    #
+    extra_packages = []
+
+    if args.bootable:
+        extra_packages += ["kernel-default"]
+
+    if args.encrypt:
+        extra_packages += ["device-mapper"]
+
+    if args.output_format in (OutputFormat.subvolume, OutputFormat.raw_btrfs):
+        extra_packages += ["btrfsprogs"]
+
+    if args.packages:
+        extra_packages += args.packages
+
+    if run_build_script and args.build_packages is not None:
+        extra_packages += args.build_packages
+
+    if extra_packages:
+        subprocess.run(cmdline + extra_packages, check=True)
+
+    #
+    # Disable packages caching in the image that was enabled
+    # previously to populate the package cache.
+    #
+    subprocess.run(["zypper", "--root", root, "modifyrepo", "-K", "Main"], check=True)
+    subprocess.run(["zypper", "--root", root, "modifyrepo", "-K", "Updates"], check=True)
```

Comment:
@poettering, I don't think it's possible to merge all calls to zypper in one call, but I'm far from a zypper expert so I can be wrong. OTOH I'm not sure why you want to limit the number of calls as it will definitively makes the code harder to parse in this case.